File: Android/Blink/libandroid/src/main/java/net/qiujuer/blink/ExecutorDelivery.java
package net.qiujuer.blink;

import android.os.Handler;

import net.qiujuer.blink.core.ReceiveDelivery;
import net.qiujuer.blink.core.ReceivePacket;
import net.qiujuer.blink.core.SendDelivery;
import net.qiujuer.blink.core.SendPacket;
import net.qiujuer.blink.listener.ReceiveListener;

import java.util.concurrent.Executor;

/**
 * Delivers send and receive responses.
 */
public class ExecutorDelivery extends ReceiveDelivery implements SendDelivery {
    /**
     * Used for posting responses, typically to the main thread.
     */
    private final Executor mPoster;

    public ExecutorDelivery(final Handler handler, ReceiveListener listener) {
        super(listener);
        mPoster = new Executor() {
            @Override
            public void execute(Runnable command) {
                handler.post(command);
            }
        };
    }

    public ExecutorDelivery(Executor executor, ReceiveListener listener) {
        super(listener);
        mPoster = executor;
    }

    @Override
    public void postSendStart(SendPacket entity) {
        mPoster.execute(new SendDeliveryRunnable(entity, null, false));
    }

    @Override
    public void postSendEnd(SendPacket entity, boolean isSuccess) {
        entity.setSuccess(isSuccess);
        mPoster.execute(new SendDeliveryRunnable(entity, null, true));
    }

    @Override
    public void postSendProgress(SendPacket entity, int total, int cur) {
        ProgressStatus status = new ProgressStatus(total, cur);
        mPoster.execute(new SendDeliveryRunnable(entity, status, false));
    }

    @Override
    public void postReceiveStart(ReceivePacket entity) {
        mPoster.execute(new ReceiveDeliveryRunnable(getReceiveListener(), entity, null, false));
    }

    @Override
    public void postReceiveEnd(ReceivePacket entity, boolean isSuccess) {
        entity.setSuccess(isSuccess);
        mPoster.execute(new ReceiveDeliveryRunnable(getReceiveListener(), entity, null, true));
    }

    @Override
    public void postReceiveProgress(ReceivePacket entity, int total, int cur) {
        ProgressStatus status = new ProgressStatus(total, cur);
        mPoster.execute(new ReceiveDeliveryRunnable(getReceiveListener(), entity, status, false));
    }

    private class ReceiveDeliveryRunnable implements Runnable {
        private ReceiveListener listener;
        private ReceivePacket entity;
        private ProgressStatus status;
        private boolean isEnd;

        public ReceiveDeliveryRunnable(ReceiveListener listener, ReceivePacket entity, ProgressStatus status, boolean isEnd) {
            this.listener = listener;
            this.entity = entity;
            this.status = status;
            this.isEnd = isEnd;
        }

        @SuppressWarnings("unchecked")
        @Override
        public void run() {
            if (listener != null) {
                if (status != null)
                    listener.onReceiveProgress(entity.getType(), entity.getId(), status.total, status.cur);
                else if (isEnd)
                    listener.onReceiveEnd(entity);
                else
                    listener.onReceiveStart(entity.getType(), entity.getId());
            }
            entity = null;
            status = null;
            listener = null;
        }
    }


    private class SendDeliveryRunnable implements Runnable {
        private SendPacket entity;
        private ProgressStatus status;
        private boolean isEnd;

        public SendDeliveryRunnable(SendPacket entity, ProgressStatus status, boolean isEnd) {
            this.entity = entity;
            this.status = status;
            this.isEnd = isEnd;
        }

        @SuppressWarnings("unchecked")
        @Override
        public void run() {
            if (!entity.isCanceled()) {
                if (status != null)
                    entity.deliverProgress(status.total, status.cur);
                else if (isEnd)
                    entity.deliverEnd();
                else
                    entity.deliverStart();
            }

            entity = null;
            status = null;
        }
    }

    /**
     * Send progress status
     */
    private class ProgressStatus {
        int total;
        int cur;

        ProgressStatus(int total, int cur) {
            this.total = total;
            this.cur = cur;
        }
    }
}

File: C Sharp/Blink/Blink/ExecutorDelivery.cs
using Net.Qiujuer.Blink.Core;
using Net.Qiujuer.Blink.Listener;
using Net.Qiujuer.Blink.Tool;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Net.Qiujuer.Blink
{
    class ExecutorDelivery : IReceiveDelivery, ISendDelivery
    {
        private event EventHandler<Runnable> mBlinkPoster;
        private event EventHandler<Runnable> mSendPoster;
        private readonly BlinkListener mBlinkListener;

        public ExecutorDelivery(BlinkListener listener)
        {
            mBlinkPoster += ExecutorDelivery_BlinkPoster;
            mSendPoster += ExecutorDelivery_SendPoster;

            mBlinkListener = listener;
        }

        void ExecutorDelivery_SendPoster(object sender, Runnable e)
        {
            e.Run();
        }

        void ExecutorDelivery_BlinkPoster(object sender, Runnable e)
        {
            e.Run();
        }

        public void PostReceiveStart(ReceivePacket entity)
        {
            mBlinkPoster.BeginInvoke(null, new ReceiveDeliveryRunnable(GetBlinkListener(), entity, 0, false), null, null);
        }

        public void PostReceiveEnd(ReceivePacket entity, bool isSuccess)
        {
            entity.SetSuccess(isSuccess);
            mBlinkPoster.BeginInvoke(null, new ReceiveDeliveryRunnable(GetBlinkListener(), entity, 0, true), null, null);
        }

        public void PostReceiveProgress(ReceivePacket entity, float progress)
        {
            mBlinkPoster.BeginInvoke(null, new ReceiveDeliveryRunnable(GetBlinkListener(), entity, progress, false), null, null);
        }

        public void PostSendStart(SendPacket entity)
        {
            mSendPoster.BeginInvoke(null, new SendDeliveryRunnable(entity, 0, false), null, null);
        }

        public void PostSendEnd(SendPacket entity, bool isSuccess)
        {
            entity.SetSuccess(isSuccess);
            mSendPoster.BeginInvoke(null, new SendDeliveryRunnable(entity, 0, true), null, null);
        }

        public void PostSendProgress(SendPacket entity, float progress)
        {
            mSendPoster.BeginInvoke(null, new SendDeliveryRunnable(entity, progress, false), null, null);
        }


        /**
         * Get ReceiveListener
         *
         * @return ReceiveListener
         */
        protected BlinkListener GetBlinkListener()
        {
            return mBlinkListener;
        }

        private class ReceiveDeliveryRunnable : Runnable
        {
            private BlinkListener listener;
            private ReceivePacket entity;
            private float progress;
            private bool isEnd;

            public ReceiveDeliveryRunnable(BlinkListener listener, ReceivePacket entity, float progress, bool isEnd)
            {
                this.listener = listener;
                this.entity = entity;
                this.progress = progress;
                this.isEnd = isEnd;
            }

            public void Run()
            {
                if (listener != null)
                {
                    if (progress > 0)
                        listener.OnReceiveProgress(entity, progress);
                    else if (isEnd)
                        listener.OnReceiveEnd(entity);
                    else
                        listener.OnReceiveStart(entity.GetType(), entity.GetId());
                }
                entity = null;
                listener = null;
            }
        }


        private class SendDeliveryRunnable : Runnable
        {
            private SendPacket entity;
            private float progress;
            private bool isEnd;

            public SendDeliveryRunnable(SendPacket entity, float progress, bool isEnd)
            {
                this.entity = entity;
                this.progress = progress;
                this.isEnd = isEnd;
            }


            public void Run()
            {
                if (!entity.IsCanceled())
                {
                    if (progress > 0)
                        entity.DeliverProgress(progress);
                    else if (isEnd)
                        entity.DeliverEnd();
                    else
                        entity.DeliverStart();
                }

                entity = null;
            }
        }
    }
}
